packages/upload-media/README.md

```diff
@@ -0,0 +1,181 @@
+-   _$0.onChange_ `[OptimizeExistingItemArgs[ 'onChange' ]]`: Function called each time a file or a temporary representation of the file is available.
+-   _$0.onSuccess_ `[OptimizeExistingItemArgs[ 'onSuccess' ]]`: Function called after the file is uploaded.
+-   _$0.onBatchSuccess_ `[OptimizeExistingItemArgs[ 'onBatchSuccess' ]]`: Function called after a batch of files is uploaded.
+-   _$0.onError_ `[OptimizeExistingItemArgs[ 'onError' ]]`: Function called when an error happens.
+-   _$0.additionalData_ `[OptimizeExistingItemArgs[ 'additionalData' ]]`: Additional data to include in the request.
+
+#### rejectApproval
+
+Rejects a proposed optimized/converted version of a file by essentially cancelling its further processing.
+
+_Parameters_
+
+-   _id_ `number`: Item ID.
+
+<!-- END TOKEN(Autogenerated actions|src/store/actions.ts) -->
+
+### Selectors
+
+The following selectors are available on the object returned by `wp.data.select( 'core/upload-media' )`:
+
+<!-- START TOKEN(Autogenerated selectors|src/store/selectors.ts) -->
+
+#### getComparisonDataForApproval
+
+Returns data to compare the old file vs. the optimized file, given the attachment ID.
+
+Includes both the URLs and the respective file sizes and the size difference in percentage.
+
+_Parameters_
+
+-   _state_ `State`: Upload state.
+-   _attachmentId_ `number`: Attachment ID.
+
+_Returns_
+
+-   `{ oldUrl: string | undefined; oldSize: number; newSize: number; newUrl: string | undefined; sizeDiff: number; } | null`: Comparison data.
+
+#### getItems
+
+Returns all items currently being uploaded, without sub-sizes (children).
+
+_Parameters_
+
+-   _state_ `State`: Upload state.
+
+_Returns_
+
+-   `QueueItem[]`: Queue items.
+
+#### getSettings
+
+Returns the media upload settings.
+
+_Parameters_
+
+-   _state_ `State`: Upload state.
+
+_Returns_
+
+-   `Settings`: Settings
+
+#### isPendingApproval
+
+Determines whether there is an item pending approval.
+
+_Parameters_
+
+-   _state_ `State`: Upload state.
+
+_Returns_
+
+-   `boolean`: Whether there is an item pending approval.
+
+#### isPendingApprovalByAttachmentId
+
+Determines whether an item is the first one pending approval given its associated attachment ID.
+
+_Parameters_
+
+-   _state_ `State`: Upload state.
+-   _attachmentId_ `number`: Attachment ID.
+
+_Returns_
+
+-   `boolean`: Whether the item is first in the list of items pending approval.
+
+#### isUploading
+
+Determines whether any upload is currently in progress.
+
+_Parameters_
+
+-   _state_ `State`: Upload state.
+
+_Returns_
+
+-   `boolean`: Whether any upload is currently in progress.
+
+#### isUploadingById
+
+Determines whether an upload is currently in progress given an attachment ID.
+
+_Parameters_
+
+-   _state_ `State`: Upload state.
+-   _attachmentId_ `number`: Attachment ID.
+
+_Returns_
+
+-   `boolean`: Whether upload is currently in progress for the given attachment.
+
+#### isUploadingByUrl
```

Comment:
Why do we need both isUploadingById and isUploadingByUrl?

Reply:
Both serve different purposes:

isUploadingByUrl: For example when importing an external image you only have a URL

isUploadingById: For example when compressing an existing attachment, you have an attachment ID so you can easily know whether there is an ongoing operation.

Reply:
I wonder if the "add/optimize" action should return some kind of identifier to be able to track this instead with a single function and a single approach. itemId or something.

Reply:
Since when can actions return something? 🤔

Reply:
I think some actions can return promises (thunks maybe) but I might be wrong. Anyway, it's not that important, I was just thinking out loud.

Reply:
Ah, they're basically all thunks, but they all return `void`, never thought about returning something from them.

The downside would be that you'd have to `await` the action to get the return value, and that this would expose the `itemId`, which is deliberately private and not exposed elsewhere (implementation detail).

Something that could be revisited in further iterations of this functionality I suppose.